File: inference/code.py
from palmerpenguins import load_penguins
import pandas as pd
import numpy as np
import altair as alt

# load and clean data
penguins = load_penguins().dropna()

# treat the full Adelie penguin flipper length data as our "population"
adelie = penguins[penguins["species"] == "Adelie"]
population = adelie["flipper_length_mm"].copy()
population_mean = population.mean()
population_size = len(population)

print(f"Population mean: {population_mean:.2f} mm")
print(f"Population size: {population_size}")

# take many samples from the population (without replacement)
np.random.seed(42)
n_samples = 1000
sample_size_n = 30  # Each sample has 30 penguins
sample_means = []

for i in range(n_samples):
    sample = population.sample(n=sample_size_n, replace=False)
    sample_means.append(sample.mean())

sampling_dist_df = pd.DataFrame({"mean_flipper_length": sample_means})

# calculate standard error
se_sampling = np.std(sample_means)
print(f"Standard error of sampling distribution: {se_sampling:.2f} mm")

# visualize sampling distribution
sampling_chart = (
    alt.Chart(sampling_dist_df)
    .mark_bar()
    .encode(
        alt.X(
            "mean_flipper_length:Q",
            bin=alt.Bin(maxbins=30),
            title="Sample Mean Flipper Length (mm)",
        ),
        alt.Y("count()", title="Count"),
    )
    .properties(
        title="Sampling Distribution of Mean Flipper Length (n=30, 1000 samples)",
        width=600,
        height=400,
    )
)

# add population mean line
pop_line = (
    alt.Chart(pd.DataFrame({"pop_mean": [population_mean]}))
    .mark_rule(color="red", size=2)
    .encode(x="pop_mean:Q")
)

sampling_chart + pop_line

File: inference/code.R
library(tidyverse)
library(infer)
library(palmerpenguins)

# -----------------------------------------------------------------------------
# Goal: Show how a sampling distribution is built by taking many samples from
# a known population. In practice we never have the full population, but this
# exercise lets us see what "repeated sampling" looks like and why sample means
# cluster around the true population mean.
# -----------------------------------------------------------------------------

# Treat the full Adelie penguin flipper length data as our "population"
population <- penguins |>
  filter(species == "Adelie", !is.na(flipper_length_mm)) |>
  select(flipper_length_mm)

population_mean <- mean(population$flipper_length_mm)
population_size <- nrow(population)

cat("Population mean:", round(population_mean, 2), "mm\n")
cat("Population size:", population_size, "\n")

# -----------------------------------------------------------------------------
# Take 1000 random samples (without replacement) from the population.
# Each sample has 30 penguins — the same size used throughout the textbook.
# rep_sample_n() adds a "replicate" column so we can group by each sample.
# -----------------------------------------------------------------------------
set.seed(42)
samples <- population |>
  rep_sample_n(size = 30, replace = FALSE, reps = 1000)

# Calculate the sample mean flipper length for each of the 1000 samples
sample_means <- samples |>
  group_by(replicate) |>
  summarize(mean_flipper_length = mean(flipper_length_mm))

# The standard error tells us how much sample means vary around the population mean
se_sampling <- sd(sample_means$mean_flipper_length)
cat("Standard error of sampling distribution:", round(se_sampling, 2), "mm\n")

# -----------------------------------------------------------------------------
# Visualize the sampling distribution.
# Each bar represents how often samples produced a mean in that range.
# The red line marks the true population mean — notice the distribution is
# centred on it. This illustrates that sample means are unbiased estimators.
# -----------------------------------------------------------------------------
sampling_plot <- ggplot(sample_means, aes(x = mean_flipper_length)) +
  geom_histogram(binwidth = 0.5, colour = "white", fill = "steelblue") +
  geom_vline(xintercept = population_mean, colour = "red", linewidth = 1) +
  labs(
    title = "Sampling Distribution of Mean Flipper Length (n = 30, 1000 samples)",
    x = "Sample Mean Flipper Length (mm)",
    y = "Count"
  ) +
  theme_minimal()

sampling_plot
